src/main/kotlin/org/taktik/icure/asynclogic/objectstorage/DataAttachmentModificationLogic.kt

```diff
@@ -0,0 +1,80 @@
+package org.taktik.icure.asynclogic.objectstorage
+
+import kotlinx.coroutines.flow.Flow
+import org.springframework.core.io.buffer.DataBuffer
+import org.taktik.icure.entities.Document
+import org.taktik.icure.entities.base.HasDataAttachments
+import org.taktik.icure.entities.embed.DataAttachment
+import org.taktik.icure.exceptions.ObjectStoreException
+
+/**
+ * Shared logic for the modification of entities which have [DataAttachment]s.
+ */
+interface DataAttachmentModificationLogic<T : HasDataAttachments<T>> {
+	/**
+	 * Verifies that all updates to an entity with data attachments are valid in regard to
+	 * the data attachment information: some changes to attachment information can only be
+	 * executed through [updateAttachments]. This prevents accidental loss of information.
+	 *
+	 * The following changes are considered invalid:
+	 * 1. The new version of an entity specifies some attachments which do not exist in the
+	 * current version.
+	 * 2. The new version changes the value of a [DataAttachment.couchDbAttachmentId] or
+	 * [DataAttachment.objectStoreAttachmentId].
+	 * 3. Any change in the [HasDataAttachments.deletedAttachments]
+	 *
+	 * In most cases if there is an invalid change this method will throw an [IllegalArgumentException],
+	 * however it is possible to specify to have a lenient behaviour for some attachments, in order to
+	 * preserve retro-compatibility. All invalid changes to attachment data that is mapped to a key
+	 * in [lenientKeys] will be simply ignored, without triggering an [IllegalArgumentException].
+	 *
+	 * @param currEntity the current value of the entity being updated
+	 * @param newEntity the new desired value for the entity
+	 * @param lenientKeys keys of attachments to exclude from the check.
+	 * @return an updated version of newEntity which does not have any invalid change.
+	 * @throws IllegalArgumentException if there are invalid changes not related to lenient keys.
+	 */
+	fun ensureValidAttachmentChanges(currEntity: T, newEntity: T, lenientKeys: Set<String>): T
+
+	/**
+	 * Updates an entity attachments, also performing any side-tasks necessary for the appropriate
+	 * storage of the attachments content.
+	 * @param currEntity the current value of the entity which needs to be updated.
+	 * @param changes the changes to apply to the entity attachments.
+	 * @return the updated entity
+	 * @throws ObjectStoreException if one or more attachments must be stored using the object
+	 * storage service but this is not possible at the moment.
+	 */
+	suspend fun updateAttachments(currEntity: T, changes: Map<String, DataAttachmentChange>): T?
```

Comment:
Actually if the object storage is not available as long as the pre-store is successful (which should always be the case if the system is configured properly) there is no problem: no exception will be thrown and the operation can still succeed. The reason of the nullability is because this has to also update the entity itself, through `LookupDAO.save`, which in the signature can return null, although I think it could be updated to always return a non-nullable entity, since in case of errors it will throw an exception...
In any case I think that in case of errors it is better to propagate the exception, rather than returning the entity, so that whoever is using the method can decide how to deal with it in the best way (for example returning a 500 or 503 error when the method is used by controllers).